Refactor value validity as an addtional decode step (#244)
minirust/minirust: spec/lang/representation.md

# MiniRust representation relation

> The language specification enforces (by penalty of Undefined Behavior) that when a range of bytes is "used at a given type",
> it satisfies an invariant -- the *language invariant* (or "validity invariant").
> This is the main purpose of [types](types.md): to define how [values](values.md) are (de)serialized into/from memory.

Reviewer: ```suggestion
The main purpose of [types](types.md) is to define how [values](values.md) are (de)serialized into/from memory.
This is the *[representation relation]*, which is defined in the following.
We also use this file to define the *language invariant* (or "validity invariant"): a predicate on byte ranges that must always hold when a range of bytes is "used at a given type".
For most types, the language invariant is defined by "these bytes represent some value", but for pointers, the language invariant imposes extra requirements.
```

> This *[representation relation]* for a type is defined in the following.
> `decode` converts a list of bytes into a (potentially not well-formed) value; this operation can already fail if the byte list is not a valid encoding for the given type.
> The values are always checked after decoding to be [well-formed][well-formed-value] by `check_value`, since some types also have additional runtime constraints.
> `encode` inverts `decode`; it will always succeed for [well-formed][well-formed-value] inputs.
> 
> `encode`, `decode` and `check_value` satisfy some [properties](#generic-properties), which the specification must ensure, i.e. violating these properties is a spec bug.
> However, they can also make some assumptions which the specification ensures:
> The types these are called with/on are all [well-formed](well-formed.md#well-formed-layouts-and-types) and sized, as unsized types cannot be represented as values.
> `encode` can also assume the value is [well-formed][well-formed-value] for the type,
> and `decode` can assume the size of the byte list matches the size of the type.

Reviewer: ```suggestion
The representation relation is defined by two functions, which are duals to each other: `decode` and `encode`.
`decode` converts a list of bytes into a (potentially not well-formed) value; this operation can fail if the byte list is not a valid representation for the given type.
The decoded values must still be checked to be [well-formed][well-formed-value] by `check_value`, since some types also have additional runtime constraints.
`encode` inverts `decode`; it will always succeed for [well-formed][well-formed-value] inputs.

`encode`, `decode` and `check_value` satisfy some [properties](#generic-properties), which the specification must ensure, i.e. violating these properties is a spec bug.
However, they can also make some assumptions which the specification ensures:
The types these are called with/on are all [well-formed](well-formed.md#well-formed-layouts-and-types) and sized, as unsized types cannot be represented as values.
`encode` can also assume the value is [well-formed][well-formed-value] for the type,
and `decode` can assume the length of the byte list matches the size of the type.
```

The main purpose of [types](types.md) is to define how [values](values.md) are (de)serialized into/from memory.
This is the *[representation relation]*, which is defined in the following.
`decode` converts a list of bytes into a value; this operation can fail if the byte list is not a valid encoding for the given type.

> Some types also have additional runtime constraints to be [valid][valid-value], which is always checked after decoding.

Reviewer: ```suggestion
Some types also have additional runtime constraints to be [well-formed][well-formed-value], which is always checked after decoding.
```

`encode` inverts `decode`; it, however, will always work when the value is [valid][valid-value] for the given type.

`encode` and `decode` must satisfy some [properties](#generic-properties), which the specification must ensure, i.e. violating these properties is a spec bug.
However, they can also make some assumptions:
The types in the representation relation are all [well-formed](well-formed.md#well-formed-layouts-and-types) and sized, as unsized types cannot be represented as values.
`encode` can also assume the value is [valid][valid-value] for the type,
and `decode` can assume the size of the byte list matches the size of the type.

[representation relation]: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/glossary.md#representation-relation

> [valid-value]: #Value-validity

Reviewer: I am not sure we want to rename "well-formed" to "valid"... I am trying to move away from the term "validity invariant" (see https://github.com/rust-lang/unsafe-code-guidelines/issues/539), so it seems a step backwards for MiniRust to start using this term now.

```rust
impl Type {
    fn decode<M: Memory>(Type::Slice { .. }: Self, bytes: List<AbstractByte<M::Provenance>>) -> Option<Value<M>> {
        panic!("decode of Type::Slice")
    }
    fn encode<M: Memory>(Type::Slice { .. }: Self, val: Value<M>) -> List<AbstractByte<M::Provenance>> {
        panic!("encode of Type::Slice")
    }
}
```

> ## Value validity

Reviewer: ```suggestion
## Language invariant for valid representations

The language specification enforces (by penalty of Undefined Behavior) that when a range of bytes is "used at a given type", it satisfies an invariant -- the *language invariant* (or "validity invariant"). [...]
```

                // Safe pointer, i.e. references, boxes
                if let Some(layout) = ptr_ty.safe_pointee() {
                    let size = layout.size.compute(ptr.metadata);
                    // The total size of slices must be at most `isize::MAX`.
                    ensure_ub(size.bytes().in_bounds(Signed, M::T::PTR_SIZE), "Value::Ptr: total size exeeds isize::MAX")?;

                    // Safe addresses need to be non-null, aligned, dereferenceable, and not point to an uninhabited type.
                    // (Think: uninhabited types have impossible alignment.)
                    ensure_ub(ptr.thin_pointer.addr != 0, "Value::Ptr: null safe pointer")?;
                    ensure_ub(layout.align.is_aligned(ptr.thin_pointer.addr), "Value::Ptr: unaligned safe pointer")?;
                    ensure_ub(layout.inhabited, "Value::Ptr: safe pointer to uninhabited type")?;
                    ensure_ub(
                        self.mem.dereferenceable(ptr.thin_pointer, size).is_ok(),
                        "Value::Ptr: non-dereferenceable safe pointer"
                    )?;

>                     // However, is it not UB, if the validity invariant of the pointee is broken.

Reviewer: ```suggestion
                    // However, we do not care about the data stored wherever this pointer points to.
```

                    _ => {}
                };
            }
            (Value::Tuple(vals), Type::Tuple { fields, .. }) => {
                ensure_ub(vals.len() == fields.len(), "Value::Tuple: invalid number of fields")?;
                for (val, (_, ty)) in vals.zip(fields) {
                    self.check_value(val, ty)?;
                }
            }
            (Value::Tuple(vals), Type::Array { elem, count }) => {
                ensure_ub(vals.len() == count, "Value::Tuple: invalid number of elements")?;
                for val in vals {
                    self.check_value(val, elem)?;
                }
            }
            (Value::Union(chunk_data), Type::Union { chunks, .. }) => {
                ensure_ub(chunk_data.len() == chunks.len(), "Value::Union: invalid chunk size")?;
                for (data, (_, size)) in chunk_data.zip(chunks) {
                    ensure_ub(data.len() == size.bytes(), "Value::Union: invalid chunk data")?;
                }
            }
            (Value::Variant { discriminant, data }, Type::Enum { variants, .. }) => {
                let Some(variant) = variants.get(discriminant) else {

>                     throw_ub!("Value::Variant: invalid discrimant type");

Reviewer: ```suggestion
                    throw_ub!("Value::Variant: invalid discrimant");
```

    fn typed_store(&mut self, ptr: ThinPointer<M::Provenance>, val: Value<M>, ty: Type, align: Align, atomicity: Atomicity) -> Result {
        // All values floating around in minirust must be valid.
        assert!(self.check_value(val, ty).is_ok(), "trying to store {val:?} which is ill-formed for {:#?}", ty);
        let bytes = ty.encode::<M>(val);
        self.mem.store(ptr, bytes, align, atomicity)?;

        ret(())
    }

    fn typed_load(&mut self, ptr: ThinPointer<M::Provenance>, ty: Type, align: Align, atomicity: Atomicity) -> Result<Value<M>> {
        let bytes = self.mem.load(ptr, ty.size::<M::T>().expect_sized("the callers ensure `ty` is sized"), align, atomicity)?;
        ret(match ty.decode::<M>(bytes) {
            Some(val) => {
                // Ensures we only produce valid values.
                self.check_value(val, ty)?;
                val
            }
            None => throw_ub!("load at type {ty:?} but the data in memory violates the validity invariant"), // FIXME use Display instead of Debug for `ty`
        })
    }
}

impl<M: Memory> ConcurrentMemory<M> {
    /// Find all pointers in this value, ensure they are valid, and retag them.

>     fn retag_val(&mut self, frame_extra: &mut M::FrameExtra, val: Value<M>, ty: Type, fn_entry: bool) -> Result<Value<M>> {

Reviewer: This should probably be moved to the `Validating a value` section in `statements.md`.

```rust
impl Type {
    fn decode<M: Memory>(Type::Slice { .. }: Self, bytes: List<AbstractByte<M::Provenance>>) -> Option<Value<M>> {
        panic!("decode of Type::Slice")
    }
    fn encode<M: Memory>(Type::Slice { .. }: Self, val: Value<M>) -> List<AbstractByte<M::Provenance>> {
        panic!("encode of Type::Slice")
    }
}
```

## Well-formed values

We call a value `val` *well-formed* for a type `ty` if `machine.check_value(val, ty).is_ok()`.

> The specification ensures (statically or dynamically) that all values in MiniRust are always well-formed, the only exception being values returned from `decode`, which are sanitized by [typed memory accesses](#Typed-memory-accesses).

Reviewer: ```suggestion
The specification ensures (statically or dynamically) that all values in MiniRust are always well-formed, the only exception being values returned from `decode`, which  must be passed to `check_value` before being passed to any other part of the specification.
```

Violating this when running a well-formed program is a spec bug.

> Therefore `encode` and other expressions can assume well-formed inputs as well.

Reviewer: ```suggestion
Therefore, `encode` and other expressions can assume well-formed inputs as well.
```

> In particular must a load instruction cause UB, when the sequence of bytes does not correspond to any representation of a well-formed value.

Reviewer: ```suggestion
In particular, a load instruction causes UB when the sequence of bytes does not correspond to a representation of a well-formed value.
```

Generally, this is already dealt with by `decode`: if decoding succeeds, the representation is well-formed.
However, pointer types have further runtime constraints: specifically, [safe pointers][ptr_type] require the address to be aligned, dereferenceable and non-null, and they require the pointee type to be inhabited.

> This is separated from `decode`, because checking alignment or the well-formedness of a trait object pointer needs vtable information, and dereferenceablity need the allocations.

Reviewer: ```suggestion
This is separated from `decode`, because checking alignment or the well-formedness of a trait object pointer needs vtable information, and dereferenceablity need information about the current state of memory.
```

Dropping dereferenceablity as a requirement to unify these steps was discussed, see [this discussion](https://github.com/rust-lang/unsafe-code-guidelines/issues/77), but for trait objects this separation is desireable.
Therefore usages for `decode` will either follow up by calling `check_value` to raise UB or reason about why this is statically ensured already. 

So types like `&!`, have many byte lists which decode without problems, but none are *well-formed*:

> when the pointee type is uninhabited (as given by the pointee type), there exists no valid reference to that type.

Reviewer: ```suggestion
when the pointee type is uninhabited, there exists no valid reference to that type.
```

[ptr_type]: ../mem/pointer.md#Pointee

```rust
/// Ensures the given boolean is true or else raises UB.
fn ensure_else_ub(b: bool, msg: &str) -> Result<()> {
    if !b { throw_ub!("{}", msg); }
    ret(())
}

impl<M: Memory> Machine<M> {
    /// We assume `ty` is itself well-formed.
    fn check_value(&self, value: Value<M>, ty: Type) -> Result {
        match (value, ty) {
            (Value::Int(i), Type::Int(int_ty)) => {
                ensure_else_ub(int_ty.can_represent(i), "Value::Int: invalid integer value")?;
            }
            (Value::Bool(_), Type::Bool) => {},
            (Value::Ptr(ptr), Type::Ptr(ptr_ty)) => {
                ensure_else_ub(ptr_ty.meta_kind().matches(ptr.metadata), "Value::Ptr: invalid metadata")?;
                ensure_else_ub(ptr.thin_pointer.addr.in_bounds(Unsigned, M::T::PTR_SIZE), "Value::Ptr: pointer out-of-bounds")?;

                // Safe pointer, i.e. references, boxes
                if let Some(layout) = ptr_ty.safe_pointee() {
                    let size = layout.size.compute(ptr.metadata);

>                     // The total size of slices must be at most `isize::MAX`.

Reviewer: ```suggestion
                    // The total size must be at most `isize::MAX`.
```

                for (val, (_, ty)) in vals.zip(fields) {
                    self.check_value(val, ty)?;
                }
            }
            (Value::Tuple(vals), Type::Array { elem, count }) => {
                ensure_else_ub(vals.len() == count, "Value::Tuple: invalid number of elements")?;
                for val in vals {
                    self.check_value(val, elem)?;
                }
            }
            (Value::Union(chunk_data), Type::Union { chunks, .. }) => {
                ensure_else_ub(chunk_data.len() == chunks.len(), "Value::Union: invalid chunk size")?;
                for (data, (_, size)) in chunk_data.zip(chunks) {
                    ensure_else_ub(data.len() == size.bytes(), "Value::Union: invalid chunk data")?;
                }
            }
            (Value::Variant { discriminant, data }, Type::Enum { variants, .. }) => {
                let Some(variant) = variants.get(discriminant) else {
                    throw_ub!("Value::Variant: invalid discrimant");
                };
                self.check_value(data, variant.ty)?;
            }

>             (_, Type::Slice { .. }) => throw_ub!("Value: slices cannot be represented as values"),

Reviewer: Shouldn't this be `panic` since such a value is not supposed to even get here?

            _ => throw_ub!("Value: value does not match type")
        }

        ret(())
    }
}
```

- TODO: Do we really want to special case references to uninhabited types? Do we somehow want to require more, like pointing to a valid instance of the pointee type?
  (The latter would not even be possible with the current pointee information in MiniRust.)
  Also see [this discussion](https://github.com/rust-lang/unsafe-code-guidelines/issues/77).

> - TODO: Inhabitedness and non-nullness are (the only two) conditions that could be checked in `decode` itself. Should we?

Reviewer: ```suggestion
- TODO: Inhabitedness, alignment, and non-nullness could be checked in `decode` itself. Should we?
```

Author: I removed alignment there, because it can only be checked for sized pointees.
